### PR TITLE
Address modes

### PR DESCRIPTION
**Pull request by SenorBlanco**
Implement sampler address modes (repeat, mirrored repeat, clamp-to-edge).
Add an end-to-end test for Samplers.

### CODE REVIEW

File: src/tests/end2end/SamplerTests.cpp [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+            .GetResult();
+
+        auto pipelineLayout = device.CreatePipelineLayoutBuilder()
+            .SetBindGroupLayout(0, mBindGroupLayout)
+            .GetResult();
+
+        auto vsModule = utils::CreateShaderModule(device, nxt::ShaderStage::Vertex, R"(
+            #version 450
+            layout(location = 0) in vec4 pos;
+            void main() {
```
Comment (Kangz) [inline src/tests/end2end/SamplerTests.cpp]: nit: to make the test even more minimal, you could have the vertices defined in the shader, and do dead code removal. See https://github.com/Kangz/nxt-standalone/blob/master/src/tests/end2end/BlendStateTests.cpp#L36

File: src/tests/end2end/SamplerTests.cpp [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+
+        auto texture = device.CreateTextureBuilder()
+            .SetDimension(nxt::TextureDimension::e2D)
+            .SetExtent(2, 2, 1)
+            .SetFormat(nxt::TextureFormat::R8G8B8A8Unorm)
+            .SetMipLevels(1)
+            .SetAllowedUsage(nxt::TextureUsageBit::TransferDst | nxt::TextureUsageBit::Sampled)
+            .GetResult();
+
+        RGBA8 data[64 * 2];
```
Comment (Kangz) [inline src/tests/end2end/SamplerTests.cpp]: There's a bunch of magic numbers here, can it be rephrased in term of kTextureRowPitchAlignment ? You should only need RGBA8 data[kTextureRowPitchAlignment / sizeof(RGBA8) + 2].

File: src/tests/end2end/SamplerTests.cpp [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+        memset(data, 128, sizeof(data));
```
Comment (Kangz) [inline src/tests/end2end/SamplerTests.cpp]: nit: sizeof(data), or make data an std::array.

File: src/tests/end2end/SamplerTests.cpp (hunk 2 above)
Comment (Kangz) [inline src/tests/end2end/SamplerTests.cpp]: nit: Also comment explaining what the created texture looks like.

File: src/tests/end2end/SamplerTests.cpp [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+        EXPECT_PIXEL_RGBA8_EQ(expectedU2, mFB.color, 2, 0);
+        EXPECT_PIXEL_RGBA8_EQ(expectedU3, mFB.color, 3, 0);
+        EXPECT_PIXEL_RGBA8_EQ(expectedV2, mFB.color, 0, 2);
+        EXPECT_PIXEL_RGBA8_EQ(expectedV3, mFB.color, 0, 3);
+        // TODO: add tests for W address mode, once NXT supports 3D textures
+    }
+
+    // Test drawing a rect with a checkerboard texture with different address modes.
+    void DoAddressModeTest() {
+        for (auto u : addressModes) {
```
Comment (Kangz) [inline src/tests/end2end/SamplerTests.cpp]: This function's body could go in the TEST_P instead.

File: src/tests/end2end/SamplerTests.cpp [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+    nxt::RenderPipeline mPipeline;
+    nxt::TextureView mTextureView;
+    nxt::Buffer mQuad;
+};
+
+TEST_P(SamplerTest, AddressMode) {
+    DoAddressModeTest();
+}
+
+NXT_INSTANTIATE_TEST(SamplerTest, D3D12Backend, MetalBackend, OpenGLBackend, VulkanBackend)
```
Comment (Kangz) [inline src/tests/end2end/SamplerTests.cpp]: This won't work on the Vulkan backend yet, can you remove it? Also did you check it work on D3D12, Metal and GL?

File: next.json [hunk 6]
```diff
@@ -701,6 +701,14 @@
             {"value": 1, "name":"linear"}
         ]
     },
+    "address mode": {
```
Comment (Kangz) [inline next.json]: note: in WebGPU we would probably have a longer name like TextureAddressMode or SamplerAddressMode.

**Comment by Kangz** [comment]
Sorry for the Travis failure, I'll take a look. It needs to be maintained once in a while because we use non-standard APT repos.

**Comment by Kangz** [comment]
The Linux build was just a flake but it looks like a couple changes are required: https://travis-ci.org/google/nxt-standalone/jobs/361800404#L1103

File: src/tests/end2end/SamplerTests.cpp (hunk 1 above)
Comment (SenorBlanco) [reply src/tests/end2end/SamplerTests.cpp]: Done.

File: src/tests/end2end/SamplerTests.cpp (hunk 3 above)
Comment (SenorBlanco) [reply src/tests/end2end/SamplerTests.cpp]: It is sizeof(data). :)

The 128 was just to show data that shouldn't be copied/sampled. No need tho, removed.

File: src/tests/end2end/SamplerTests.cpp (hunk 2 above)
Comment (SenorBlanco) [reply src/tests/end2end/SamplerTests.cpp]: Done.

Comment (SenorBlanco) [reply src/tests/end2end/SamplerTests.cpp]: Done.

File: src/tests/end2end/SamplerTests.cpp (hunk 4 above)
Comment (SenorBlanco) [reply src/tests/end2end/SamplerTests.cpp]: Done.

File: next.json (hunk 6 above)
Comment (SenorBlanco) [reply next.json]: Could do, but shouldn't FilterMode be TextureFilterMode, then?

**Comment by SenorBlanco** [comment]
Didn't get those locally. Maybe my config is weird.

Fixed, anyway.

File: src/tests/end2end/SamplerTests.cpp (hunk 5 above)
Comment (SenorBlanco) [reply src/tests/end2end/SamplerTests.cpp]: Removed.

Tested on all backends; passes.

File: next.json (hunk 6 above)
Comment (Kangz) [reply next.json]: Or SamplerFilterMode, yeah let's not do that in NXT.

**Comment by SenorBlanco** [comment]
Formatting fixed. AFAICT, the two remaining failures are Travis issues, unrelated to this change.

**Comment by SenorBlanco** [comment]
Squashed to two commits (implementation, test).
Rebased to tip-of-tree.

**Comment by Kangz** [comment]
Can we have one more rebase one this? Sorry for the hassle.

**Comment by Kangz** [comment]
Thanks, landing now.

**Pull request merged**